### PR TITLE
Improve memory text to video

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR allows to run https://huggingface.co/cerspense/zeroscope_v2_XL with ~12 GB VRAM by adding **feed-forward-chunking** (see [here](https://huggingface.co/blog/reformer#chunked-feed-forward-layer-in-reformer)).

This can be very useful for extremely memory-intensive use cases such as video-to-video and text-to-video but could in the future also make sense for text-to-image.

With this PR I can generate **36 frames** with https://huggingface.co/cerspense/zeroscope_v2_XL using less than 13GB aof VRAM.

The following works on a 50% fraction of a 24GB RTX4090:

```py
import torch
from diffusers import DiffusionPipeline, DPMSolverMultistepScheduler
from diffusers.utils import export_to_video
from PIL import Image

# Make sure CUDA has < 13GB VRAM
torch.cuda.set_per_process_memory_fraction(0.5)

pipe = DiffusionPipeline.from_pretrained("cerspense/zeroscope_v2_576w", torch_dtype=torch.float16)
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()
pipe.enable_vae_slicing()

prompt = "Darth Vader is surfing on waves"
video_frames = pipe(prompt, num_inference_steps=40, height=320, width=576, num_frames=36).frames
video_path = export_to_video(video_frames, output_video_path="/home/patrick/videos/video_576_darth_vader_36.mp4")

pipe = DiffusionPipeline.from_pretrained("cerspense/zeroscope_v2_XL", torch_dtype=torch.float16)
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()
pipe.enable_vae_slicing()

video = [Image.fromarray(frame).resize((1024, 576)) for frame in video_frames]

video_frames = pipe(prompt, video=video, strength=0.6).frames
video_path = export_to_video(video_frames, output_video_path="/home/patrick/videos/video_1024_darth_vader_36.mp4")
```

cc @hysts

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -161,6 +161,7 @@ def __init__(
                 AttnProcessor2_0() if hasattr(F, "scaled_dot_product_attention") and self.scale_qk else AttnProcessor()
             )
         self.set_processor(processor)
+
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:164]:
```suggestion
```

File: src/diffusers/models/attention.py [hunk 2]
```diff
@@ -119,14 +119,22 @@ def __init__(
...
+        self._chunk_size = chunk_size
+        self._chunk_dim = dim
+
     def forward(
         self,
         hidden_states: torch.FloatTensor,
         attention_mask: Optional[torch.FloatTensor] = None,
         encoder_hidden_states: Optional[torch.FloatTensor] = None,
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
         timestep: Optional[torch.LongTensor] = None,
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention.py:137]:
```suggestion
        timestep: Optional[torch.LongTensor] = None,
        cross_attention_kwargs: Dict[str, Any] = None,
```

File: tests/models/test_models_unet_3d_condition.py [hunk 3]
```diff
@@ -399,5 +399,23 @@ def test_lora_xformers_on_off(self):
...
+        model.to(torch_device)
+        model.eval()
+
+        with torch.no_grad():
+            output = model(**inputs_dict)[0]
+
+        model.enable_forward_chunking()
+        with torch.no_grad():
+            output_2 = model(**inputs_dict)[0]
+
```
Comment (pcuenca) [inline tests/models/test_models_unet_3d_condition.py:416]: Maybe add another test for `dim=1` too to be ultra sure, as it's the default for the `TextToVideoSDPipeline` pipeline

File: src/diffusers/models/attention.py [hunk 4]
```diff
@@ -171,7 +181,20 @@ def forward(
         if self.use_ada_layer_norm_zero:
             norm_hidden_states = norm_hidden_states * (1 + scale_mlp[:, None]) + shift_mlp[:, None]
 
-        ff_output = self.ff(norm_hidden_states)
+        if self._chunk_size is not None:
+            # "feed_forward_chunk_size" can be used to save memory
+            if norm_hidden_states.shape[self._chunk_dim] % self._chunk_size != 0:
+                raise ValueError(
+                    f"`hidden_states` dimension to be chunked: {norm_hidden_states.shape[self._chunk_dim]} has to be divisible by chunk size: {self._chunk_size}. Make sure to change `self._chunk_size` accordingly by calling `unet.enable_forward_chunking`."
```
Comment (pcuenca) [inline src/diffusers/models/attention.py:188]:
```suggestion
                    f"`hidden_states` dimension to be chunked: {norm_hidden_states.shape[self._chunk_dim]} has to be divisible by chunk size: {self._chunk_size}. Make sure to set an appropriate `chunk_size` when calling `unet.enable_forward_chunking`."
```

(Just to not refer to the private property)

**Review by pcuenca — APPROVED** [review]
Looks great!

File: src/diffusers/models/attention.py [hunk 5]
```diff
@@ -119,6 +119,15 @@ def __init__(
         self.norm3 = nn.LayerNorm(dim, elementwise_affine=norm_elementwise_affine)
         self.ff = FeedForward(dim, dropout=dropout, activation_fn=activation_fn, final_dropout=final_dropout)
 
+        # let chunk size default to None
+        self._chunk_size = None
+        self._chuck_dim = 0
```
Comment (sayakpaul) [inline src/diffusers/models/attention.py:124]:
```suggestion
        self._chunk_dim = 0
```

File: src/diffusers/models/attention.py [hunk 6]
```diff
@@ -119,6 +119,15 @@ def __init__(
...
+
+    def set_chunk_feed_forward(self, chunk_size: Optional[int], dim: int):
```
Comment (sayakpaul) [inline src/diffusers/models/attention.py:126]: Why not use `chunk_dim` instead of `dim`?

File: src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py [hunk 7]
```diff
@@ -634,6 +634,9 @@ def __call__(
         # 6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
         extra_step_kwargs = self.prepare_extra_step_kwargs(generator, eta)
 
+        # 6.1 Chunk feed-forward computation to save memory
+        self.unet.enable_forward_chunking(chunk_size=1, dim=1)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py:638]: Are these reasonable defaults or should we also the users to specify these too?

File: tests/models/test_models_unet_3d_condition.py (hunk 3 above)
Comment (sayakpaul) [reply tests/models/test_models_unet_3d_condition.py:416]: I would also maybe test for another compatible chunk size to ensure feature robustness.

**Review by sayakpaul — APPROVED** [review]
Great stuff, Patrick! 

Bookmarked your blog post for a weekend read!

**Comment by pcuenca** [comment]
> Bookmarked your blog post for a weekend read!

haha, same

File: src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/text_to_video_synthesis/pipeline_text_to_video_synth.py:638]: Reasonable defaults, here this means that we run one forward pass per frame

**Comment by RafailFridman** [comment]
I'm not 100% sure that it is related to this PR, but I think that after adding the line
`self.unet.enable_forward_chunking(chunk_size=1, dim=1)`
the performance dropped by ~3x compared to diffusers 0.17.1

TextToVideoPipeline is much slower when building diffusers from main compared to 0.17.1 release

**Comment by patrickvonplaten** [comment]
Thanks fro the report @RafailFridman - checking!
